File: service_py/service_py/server_diy.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

import rclpy
from rclpy.node import Node

from diy_interface.srv import QuestionAndAnswer

class ServiceServer(Node):
    def __init__(self):
        super().__init__('service_server')
        self._service = self.create_service(QuestionAndAnswer, "question_and_answer", self.server_callback)

    def server_callback(self, request, response):
        response.answer = "feel good!"
        self.get_logger().info("[python server] receive: {0}, return: {1}".format(
                                request.question, response.answer))
        return response

def main(args=None):
    rclpy.init(args=args)

    server_node = ServiceServer()
    rclpy.spin(server_node)
    
    server_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: service_cpp/src/server_diy.cpp
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "diy_interface/srv/question_and_answer.hpp"

class ServiceServer : public rclcpp::Node 
{
public:
    ServiceServer() : Node("service_server") {
        server_ = this->create_service<diy_interface::srv::QuestionAndAnswer>("question_and_answer", 
            std::bind(&ServiceServer::answer_callback, this, std::placeholders::_1, std::placeholders::_2));
    }
private:
    void answer_callback(const std::shared_ptr<diy_interface::srv::QuestionAndAnswer::Request> request,
                    std::shared_ptr<diy_interface::srv::QuestionAndAnswer::Response> response) {
        response->answer = "I'm fine, thank you.";
        RCLCPP_INFO(this->get_logger(), "[cpp server] receive: %s, return: %s", 
                                        request->question.c_str(), response->answer.c_str());
    }

private:
    rclcpp::Service<diy_interface::srv::QuestionAndAnswer>::SharedPtr server_;
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ServiceServer>());
    rclcpp::shutdown();

    return 0;
}